docs/DebuggingTheCompiler.md

```diff
@@ -553,6 +553,95 @@ the pipeline. Here is a quick summary of the various options:
   When printing IR for functions for print-[before|after]-all options, Only
   print the IR for functions whose name is in this comma separated list.
 
+## Debugging assembly and object code
+
+Understanding layout of compiler-generated metadata
+can sometimes involve looking at assembly and object code.
+
+### Working with a single file
+
+Here's how to generate assembly or object code:
+
+```
+# Emit assembly in Intel syntax (AT&T syntax is the default)
+swiftc tmp.swift -emit-assembly -Xllvm -x86-asm-syntax=intel -o tmp.S
+
+# Emit object code
+swiftc tmp.swift -emit-object -o tmp.o
+```
+
+Understanding mangled names can be hard though: `swift demangle` to the rescue!
+
+```
+swiftc tmp.swift -emit-assembly -Xllvm -x86-asm-syntax=intel -o - \
+  | swift demangle > tmp-demangled.S
+
+swiftc tmp.swift -emit-object -o tmp.o
+
+# Look at where different symbols are located, sorting by address (-n)
+# and displaying section names (-m)
+nm -n -m tmp.o | swift demangle > tmp.txt
+
+# Inspect disassembly of an existing dylib (AT&T syntax is the default)
+objdump -d --x86-asm-syntax=intel /path/to/libcake.dylib \
+  | swift demangle > libcake.S
```

Comment:
I vote for recommending `-macho` or simply `-m`, the output is much better for macho/Darwin anyway. Specifically it includes comments like:

```
1000032b4:     e8 35 09 00 00  call    0x100003bee ## symbol stub for: _$s12_Concurrency9MainActorCMa
```

Reply:
One of us (or someone) should wire up libswiftDemangle into `llvm-nm` because it has a `--demangle` flag, but that doesn't seem to work on swift symbols.

Reply:
Nice, I will add the `-macho` recommendation. 👍🏼

Hmm, when I do `man nm`, it doesn't have a `--demangle` flag. Maybe it's just undocumented? It certainly seems to work on C++ code. 🤔

Reply:
For the libswiftDemangle idea, I think it would be nice-to-have, but I'm not sure if the benefit is worth the implementation effort (maybe I'm overestimating the complexity of implementation though). I think the current workflow is good enough for my use case.

Reply:
the `--demangle` flag is for `objdump`